### PR TITLE
Add community server filter and icons, remove DDNet and KoG tabs

### LINKED ISSUE #5654: Rename Kog Tab to Community and add trusted servers
The current situation is unfair to good faith server owners whose servers don't have the checkmark, a more healthy way is to provide a way to apply to have the checkmark or be added to this new tab so they have easier discoverability and users are not scared to join.

One example is unique servers: https://uniqueclan.net/

### PR DESCRIPTION
**Pull request by Robyt3**
Remove the DDNet and KoG tabs and replace them with a community filter, which works like the country and type filters. Community information (ID, name, icon SHA256, countries, types, servers, ranked maps) is loaded dynamically from the DDNet info json.

The official checkmark icon is replaced with community icons. The icons are downloaded to the `communityicons` folder in the config directory from a URL specified in the DDNet info json. The icons are only re-downloaded when their SHA256 has changed.

Per default, the Internet tab with all communities is selected. When starting the client for the first time, only the DDNet community is selected. The community, country and type filters can also be used on the Favorites tab now. The LAN tab remains unchanged.

The country and type filters are refactored. Previously, using these filters changed the total number of servers and players that were displayed, as the filter was implemented as a separate step. Now the community, country and type filters work like the other browser filters. This also has the benefit that the server list and DDNet info json are not reloaded each time these filters are changed, as updating the local filtering/sorting is enough now.

The check for finished maps is made more efficient by using an `std::unordered_set` instead of linear search.

Closes #5654.

Screenshots:
- Before: screenshot_2023-10-08_13-12-13
- After: screenshot_2023-10-08_13-12-28

## Checklist

- [X] Tested the change ingame
- [X] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by Kaffeine** [comment]
Is it OK that you filter out all communities except DDNet by default? I thought there is a problem with the masterserver or this branch uses a custom `info` URL.
It was not really intuitive to find the filter tab and toggle the extra communities.

IOW after update, the server browser looks like this:
image

You have `br_filter_exclude_communities *ddnet` which disables all the other communities.

**Comment by Robyt3** [comment]
Yeah, it's intentional that only DDNet is shown by default, since that's like the previous behavior, where the client initially started on the DDNet tab. I'm open to changing it though, if others think we should show all servers by default. Then we could get rid of the somewhat hacky `*ddnet` default filter value.

**Comment by Kaffeine** [comment]
IMO this is a super mods unfriendly behavior.
I didn't know that DDNet tab is opened by default (probably I use the client longer than the tab exists). I had `Internet` tab open and never used/opened mod-specific tabs (DDNet/KoG). Yet the tabs are noticeable so a new player could see the `Internet` tab. IOW there was a single click to see all teeworlds servers.

On the other hand now the Community filter is somewhere inside the filters page of UiToolBox. It is much harder to find it there, and I don't see an obvious way to disable the filter entirely (in one or two clicks). It seems that now we have to enable the "communities" one by one. A new community won't be visible until a player check the filter for new items and explicitly enable it.

**Comment by Jupeyy** [comment]
I tested this change and have to say it's a huge step backwards in useability.

I never really had to touch filters, and by default i usually use the server info tab(the server list)

Switching between KoG and DDNet is extremly hard now. I barely used the Internet tab (used Favorite Tab instead)

I once proposed that we could instead allow removing and adding tabs (to do basically the same as this)

Before merging we should at least do a community poll or smth.. I'd assume many ppl dislike this

**Review by def- — CHANGES_REQUESTED** [review]
Huge thanks for implementing this. When this gets released we should split out noby from DDNet and allow other server hosters to sign up.

Set the DDNet filter only if g_Config.ClShowWelcome is 1, that should make sure that we don't change the Internet tab for existing players, and still have DDNet as default for new players.

> Edit: An even more general purpose solution would be to allow adding tabs that have user defined filters

That would be cool too, to save a specific search as a tab. But should be separate from this PR. Also there isn't much space for tabs without having them scroll.

What's missing in my view is filtering out a prefix from the server name. I'd like for DDNet and KoG servers not to show DDNet/KoG text since they have the logo already. I guess this would interact badly with search though.

**Comment by Jupeyy** [comment]
> Set the DDNet filter only if g_Config.ClShowWelcome is 1, that should make sure that we don't change the Internet tab for existing players, and still have DDNet as default for new players.

I have to say I'm a bit confused about what exactly the plan is.

From what I remember I had a big discussion with @heinrich5991 on discord, that I dislike that we promote other (closed source communities). His proposal was that a Internet only tab could make mods more interesting/visible again. Besides I'm not sure about this, and I proposed, that DDNet could still host open source mods in their DDNet group with a higher prio in the server list, I get very huge vanilla client vibes.

__"Vanilla only game modes" - Filter, which was active default in 0.6.x.__
New players only see DDNet and probably won't find the filter so easily, similar to how it was in the vanilla client once.
The linked issue also doesn't clearify this question.. So what exactly is the goal of this change?

This change certainly makes "DDraceNetwork" be what it is advertised for on steam, but i'd defs not say this is any kind of mod friendly, probably more the opposite.

**Comment by def-** [comment]
This removes the special status from KoG servers and allows other communities to be considered the same way.

**Comment by Jupeyy** [comment]
> That would be cool too, to save a specific search as a tab. But should be separate from this PR. Also there isn't much space for tabs without having them scroll.

for the tab size problem:
I'd move the globe icon to the left, iconize the default tabs: (mid icon should be smth like 🖧)

image


I'd remove the "Demos" button in this picture completely, and leave the tabs instead:

image


That gives more than enough space IMO

**Comment by def-** [comment]
> New players only see DDNet and probably won't find the filter so easily, similar to how it was in the vanilla client once.

This is by design and has always been this way in DDNet.

**Comment by Marmare314** [comment]
I agree with @Jupeyy that this change is not improving usability. I also use the KoG/DDNet tabs frequently and having to select DDNet and deselect KoG everytime is not very nice. But I also agree that custom filter tabs are a solution to this problem. The DDNet tab would then just be a pre-defined custom tab and everybody is free to add a KoG tab if they want (or any other community server for that matter). 

One thing I don't like about this is that the checkmark is removed. In my opinion it can still exist with this change. That would allow very low standards to apply for custom icons (so that basically every server can provide one), but official servers and open source mods (or whatever criteria there currently are) are still highlighted as such.

I also don't think any change to the server browser is going to improve discoverability of mods. Maybe for that there could be a game-mode of the week that is highlighted in the browser somehow.

**Comment by Robyt3** [comment]
> Set the DDNet filter only if g_Config.ClShowWelcome is 1, that should make sure that we don't change the Internet tab for existing players, and still have DDNet as default for new players.

Done.

**Comment by sjrc6** [comment]
I think it is quite a bad decision to make only ddnet servers visible by default.

**Comment by def-** [comment]
> I think it is quite a bad decision to make only ddnet servers visible by default.

Noted, but it's been this way since 2013.

**Comment by Learath2** [comment]
A certain proverb comes to mind perusing through the comments, "Give them a hand, and they'll take the whole arm." 🙃

It might be better to have the filter behave as radio options rather than checkboxes. (Clicking any should deselect all others). That reduces the number of clicks to go between communities back to 1. Ctrl+click to select multiple communities should be intuitive enough from web and file browsers. It's also less likely to be used so being hidden behind a modifier isn't that critical. Maybe a clear button to deselect all.

Adding more tabs, I just don't see a nice way to fit those without scrolling, and that's ugly. If the filters on the RHS are UX unfriendly, how about moving it to a horizontal bar right above the sort columns below the tabs? A more prominent position should make it more palatable, I hope.

Allowing custom tabs with the ability to write your own filter sounds great for powerusers, horrible UX for the common user. If the user we are trying to accommodate is incapable of finding the community filter in the right-hand bar, they will most certainly not be able to craft a custom filter. I think tabs that are not autogenerated are a no-go.

Another interesting solution might be to have collapsible communities, though that would make having a community a disadvantage for visibility in the browser, which sounds like an oxymoron.

This might be a decent candidate for polling the community during RC. I doubt just showing them screenshots will give a good feeling of the UX involved.

**Comment by sjrc6** [comment]
> > I think it is quite a bad decision to make only ddnet servers visible by default.
> 
> Noted, but it's been this way since 2013.

Cmon you know I meant conditional on the changes made in this PR (removing the internet tab)

**Comment by def-** [comment]
I don't see the difference. Previously new players see DDNet servers only, now new players see DDNet servers only. DDNet servers are part of "Internet", so I think it's fine to have the selection there instead of in separate tabs.

**Comment by Kaffeine** [comment]
@def-, before this change, a player could see mods in one obvious click on the first 'Internet' tab. Now it is much more complicated.

The difference is as big as sounds: "click the first tab on the main screen" and "toggle items in a listview of a filter tabview in a tab".

**Comment by def-** [comment]
We had reports from players who never knew about the Internet tab, so I'm not sure if there is much of a difference.

**Comment by heinrich5991** [comment]
> The difference is as big as sounds: "open the first tab on the main screen" and "toggle items in a listview of a filter tabview in a tab". From UX PoV it is a totally different level of availability.

The tab is visible by default. Then it reduces to "click DDNet on the right side of the screen".

**Comment by Marmare314** [comment]
When you select `None` as a community you can't filter the game type anymore.

**Comment by 0xfaulty** [comment]
A few more thoughts:
-We can't remove ddnet and kog from the 'Internet' tab because it's convenient to see all servers in one place
-We can't remove 'ddnet' and 'kog' tabs so that players don't have to check the verification tick box (or community icon) every time not to misclick and not to enter a fake server, because getting points is an important part of the game and also account token should not leak.

**Comment by 0xfaulty** [comment]
I thought about it and thought that nevertheless not to be toxic, I'm willing to try the version Robyt made, as I respect his work and that he decided to take on this task, in any case there will be some new feedback afterwards. 🙂

**Review by def- — APPROVED** [review]

**Comment by def-** [comment]
@Robyt3 Is it ok for you if we leave this for 17.5 and not 17.4?

**Comment by Robyt3** [comment]
> @Robyt3 Is it ok for you if we leave this for 17.5 and not 17.4?

Yeah, that's fine. We can gather some more feedback from nightly that way and maybe add some more communities as well.

**Comment by l-ouis** [comment]
Not sure if I'm a big fan, a couple things come to mind personally:

- the UI is now uglier and different icons clash (Why not just have the original blue tick for DDNet servers and maybe a yellow/green one for "likely safe/verified"?)
- how would mod communities/creators get their icon verified in this sense, and what would the cutoff be for adding or removing an icon to the list? (or maybe i'm not understanding the change)

**Comment by def-** [comment]
> the UI is now uglier and different icons clash (Why not just have the original blue tick for DDNet servers and maybe a yellow/green one for "likely safe/verified"?)

People would complain that we give preferential treatment to DDNet.

> how would mod communities/creators get their icon verified in this sense, and what would the cutoff be for adding or removing an icon to the list? (or maybe i'm not understanding the change)

So far it's a manual process of pinging me.

**Comment by Avolicious** [comment]
Nice change @Robyt3 

Hows the process to get listed as a community? What are the requirements?

**Comment by heinrich5991** [comment]
https://github.com/ddnet/ddnet/issues/7448#issuecomment-1808654835

No officially stated requirements AFAIK, just ask for it.

**Comment by l-ouis** [comment]
> People would complain that we give preferential treatment to DDNet.

I think that's fine, right now we already kind of give preferential treatment to kog.

And regardless I think there will be people complaining about not getting their gamemode accepted, etc. either way

**Comment by heinrich5991** [comment]
> > People would complain that we give preferential treatment to DDNet.
> 
> I think that's fine, right now we already kind of give preferential treatment to kog.

I think that's not fine.

**Comment by l-ouis** [comment]
Sorry let me rephrase

Right now we already give preferential treatment to kog which I agree is not fine, but I don't see the harm in giving a slight preferential treatment to ddnet and grouping all other popular community servers in another color of tick.

I don't think the best solution is letting everyone submit their own logo and stuffing them in yet another UI box, and I still have some concerns:

- How will the community names+icons be ordered?
- Let's say TW grows in popularity and we now have 20+ unique server networks that deserve an icon- will it scale well inside the communities box?
- Are there going to be regulations for community icons? (e.g. would someone be able to upload a big yellow arrow pointing to the right for their gamemode)

**Comment by heinrich5991** [comment]
> Right now we already give preferential treatment to kog which I agree is not fine, but I don't see the harm in giving a slight preferential treatment to ddnet and grouping all other popular community servers in another color of tick.

I see harm in giving DDNet preferential treatment for existing players. It means that other communities cannot compete with DDNet on merit. I think we should embrace being an open community, welcoming other game modes and servers.

**Comment by Jupeyy** [comment]
- After that logic ddnet should be renamed to teeworlds2 (purely from naming this would fit better)
- the UI is worse, ppl that don't play this game cannot judge on this and also should not! (ask the community before merging such changes please)
- the filter is still less intuitive as Internet tab, so right now this indeed is actually giving DDNet (& Kog) an even bigger advantage over normal servers
- the number of players is still very dominant in the server browser, so smaller mods still basically have no chance (e.g. vanilla mods often cap players to 16 players)
- DDNet collects donations to keep the **open source** game alive. Arguments like "ppl complain about DDNet's dominance" are fundamentally flawed. I am all for more mod support. But these mods should be open source, that's what ppl donated to, willingly or unwillingly. Closed source goes against our soul. Sadly nobody here has ever given any argument against these :/

**Comment by heinrich5991** [comment]
> I am all for more mod support. But these mods should be open source, that's what ppl donated to, willingly or unwillingly.

Consider e.g. unique servers, which are open-source and didn't have a chance to have a community before this change.

I also like open-source servers better. Although it's a bit hard to enforce. Perhaps we could convince KoG to open up their server code and then ask every new community to do the same?

**Comment by l-ouis** [comment]
> Consider e.g. unique servers, which are open-source and didn't have a chance to have a community before this change.

I think that's sort of an uninformed statement. Unique has had a small community for a long time now and I don't think this change would make it bigger by any significant amount. If anything, their community is smaller just because it's a more 'specific' gamemode compared to other race gamemodes

**Comment by Jupeyy** [comment]
> Consider e.g. unique servers, which are open-source and didn't have a chance to have a community before this change.

Ok i take your argument, then please answer me this simple question:
Would you rather find unique servers with a unfiltered internet tab, or if we have 3 tabs:
- Official (DDNet)
- Trusted (KoG, noby etc.)
- Rest (all servers except the above two)


My point in this case would be: Tabs are more visible to a non experienced person than this community filter.
The filter has many options and is placed in an "undesirable" position.

I simply don't understand that the current changes are an argument for more openess. to me it's more like the opposite tbh

**Comment by heinrich5991** [comment]
> > Consider e.g. unique servers, which are open-source and didn't have a chance to have a community before this change.
> 
> I think that's sort of an uninformed statement. Unique has had a small community for a long time now and I don't think this change would make it bigger by any significant amount. If anything, their community is smaller just because it's a more 'specific' gamemode compared to other race gamemodes

I meant community as in an extra tab/trusted sign. Of course they were able to have a community as in players before.

**Comment by heinrich5991** [comment]
> Ok i take your argument, then please answer me this simple question:
> Would you rather find unique servers with a unfiltered internet tab, or if we have 3 tabs:

Probably with the system as in this PR, and not with the three tabs.

**Comment by Jupeyy** [comment]
> Probably with the system as in this PR, and not with the three tabs.

A little demonstration:

full:
image

rest tab:
image

Note the scrollbar.
